mvcc/key_index.go

```diff
@@ -210,18 +223,21 @@ func (ki *keyIndex) compact(atRev int64, available map[revision]struct{}) {
 
 	if !g.isEmpty() {
 		n := g.walk(f)
-		// remove the previous contents.
-		if n != -1 {
-			g.revs = g.revs[n:]
-		}
 		// remove any tombstone
-		if len(g.revs) == 1 && i != len(ki.generations)-1 {
+		if len(g.revs)-1 == n && i != len(ki.generations)-1 {
```

Comment:
The unit tests calls Compact before Keep where Compaction on tombstone clears out the generation. Then the subsequent Keep on the same tombstone rev will not even hit `if len(g.revs)-1 == n && i != len(ki.generations)-1 {` since `if !g.isEmpty() {` will be empty.

e.g
```go
// Init Tree.
// key: "foo"
// rev: 16
// generations:
//    {empty}
//    {{14, 0}[1], {14, 1}[2], {16, 0}(t)[3]}
//    {{8, 0}[1], {10, 0}[2], {12, 0}(t)[3]}
//    {{2, 0}[1], {4, 0}[2], {6, 0}(t)[3]}

// Call Compact(12) keep = {}
// generations:
//    {empty}
//    {{14, 0}[1], {14, 1}[2], {16, 0}(t)[3]}

// Call Keep(12) keep = {} since there is no generation that contains rev 12.
```